core/build/cli.odin

```diff
@@ -0,0 +1,182 @@
+				case 1: // only key, no value
+					flag_arg.key = equal_slice[0]
+				case 2: /// key and value
+					flag_arg.key = equal_slice[0]
+					flag_arg.val = equal_slice[1]
+				}
+			
+			case: // more than 1 colon found. Invalid syntax
+				err = .Invalid_Format
+				return
+			}
+			args[i] = flag_arg
+		} else {
+			args[i] = os_arg
+		}
+	}
+	return
+}
+
+
+Flag_Desc :: struct {
+	mode: Run_Mode,
+	flag: Flag_Arg,
+	help: string,
+}
+
+Cli_Info :: struct {
+	project: ^Project, // this can be projects: []^Project
+	default_target: ^Target,
+}
+
+Cli_Flags_Error :: enum {
+	Incompatible_Flags,
+}
+
+builtin_flags := [?]Flag_Desc {
+	{.Help, {"-help", "", ""}, "Displays information about the build system or the target specified."},
+	{.Dev, {"-ols", "", ""}, "Generates an ols.json for the configuration."},
+	{.Dev, {"-vscode", "", ""}, "Generates .vscode folder for debugging."},
+	{.Dev, {"-build-pre-launch", "", ""}, "Runs the build system before debugging. (WIP)"},
+	{.Dev, {"-include-build-system", "<args>", ""}, "Include the build system as a debugging target. (WIP)"},
+	{.Dev, {"-cwd", "<dir>", ""}, "Sets the CWD to the specified directory."},
+	{.Dev, {"-cwd-workspace", "", ""}, "Sets the CWD to the root of the build system executable."},
+	{.Dev, {"-cwd-out", "", ""}, "Sets the CWD to the output directory specified in the -out odin flag"},
+	{.Dev, {"-launch-args", "<args>", ""}, "The arguments to be sent to the output executable when debugging."},
+	{.Dev, {"-dbg", "<debugger name>", ""}, "Debugger type used. Works with -vscode. Sets the ./vscode/launch.json \"type\" argument"},
+}
+
+print_general_help :: proc(info: Cli_Info) {
+	fmt.printf("Syntax: %s <flags> <target>\n", filepath.base(os.args[0]))
+	fmt.printf("Available Targets:\n")
+	for target in info.project.targets {
+		prefixed_name := strings.concatenate({info.project.target_prefix, target.name}, context.temp_allocator)
+		fmt.printf("\t%s\n", prefixed_name)
+	}
+	fmt.println()
+	fmt.printf("Builtin Flags - Only 1 [Type] group per call. Groups are incompatible\n")
+	
+	for flag_desc in builtin_flags {
+		fmt.printf("\t%s", flag_desc.flag.flag)
+		if flag_desc.flag.key != "" {
+			fmt.printf(":\"%s\"", flag_desc.flag.key)
+		}
+		if flag_desc.flag.val != "" {
+			fmt.printf("=\"%s\"", flag_desc.flag.key)
+		}
+		fmt.println()
+		fmt.printf("\t\t%s %s\n", mode_strings[flag_desc.mode], flag_desc.help)
+		fmt.println()
+	}
+}
+
+run_cli :: proc(info: Cli_Info, cli_args: []string) -> bool {
+	runtime.DEFAULT_TEMP_ALLOCATOR_TEMP_GUARD()
+	args, err := parse_args(cli_args, context.temp_allocator) // cli_args[0] should be the executable
+	target_names_from_args := make_dynamic_array_len_cap([dynamic]string, 0, len(info.project.targets), context.temp_allocator) // This is not fully correct right now
+	filtered_args_by_mode := make([dynamic]Arg, context.temp_allocator)
+	current_mode: Maybe(Run_Mode)
+	last_flag: Flag_Arg
+	for arg in args {
+		#partial switch v in arg {
```

Comment:
Why is this `#partial`?

Reply:
will fix. Artifact from when i tried to have the flags be parsed to data types like ints and bools